### ISSUE TITLE
[Bug]: Problems with columns of checkbox type with a validator

### ISSUE THREAD

**Issue by emadebManyDesigns** [handsontable/handsontable#10573]
### Describe the bug

There is a bug with checkbox columns with a validator associated: for example
https://jsfiddle.net/emadeb/mzL6yjvx/18/

When the validator is disabled, the checkbox columns works without problem: 
you can click the checkbox as fast as you want, and the checkbox is correctly changed.
If you click quicly two times on a false checkbox, the value is changed to true, than to false, as expected.

At soon as a validator is added (in my example I've added a simple validator that return always true), the checkbox doesn't work correctly: 
sometimes (in this example you need to click quickly, but in my application with a lot of data and a lot of render it appens even with a single click) the checkbox value is restored from handsontable to the original value.
For example if you click quickly two times on a false checkbox, the value is changed to true, than again to true!!!
The second click should have changed the value to false.

The problem is created by:
- the checkbox click call the `this.setDataAtCell = function(row, column, value, source, applyChangesCallback)` in the core.js: here the validation is called before the applyChanges
```
      validateChanges(processedChanges, changeSource, isValid => {
        applyChanges(processedChanges, changeSource);
        if (applyChangesCallback) applyChangesCallback(isValid, true);
      });
```
- the problem is that `this.validateCell = function(value, cellProperties, callback, source)` function of core.js call `instance.view._wt.getSetting('cellRenderer', renderableRow, renderableColumn, td);`
- this call the `cellRenderer: (renderedRowIndex, renderedColumnIndex, TD)` of tableView.js, that get the checkbox value (using the call `let value = this.instance.getDataAtRowProp(visualRowToCheck, prop);`), that return the old value, because the applyChanges is not yet called. The render is called with the old value, so the handsontable grid is updated with the old value, instead that the new one:
```
        this.instance.getCellRenderer(cellProperties)(
          this.instance,
          TD,
          visualRowIndex,
          visualColumnIndex,
          prop,
          value,
          cellProperties
        );
```

I still have to find a custom fix (waiting for the official fix), so for now I can't help you.
Surely the validateChanges should avoid to update the checkbox with the old value before the applyChanges is called.

### Provide a link to the demo with the bug reproduction

https://jsfiddle.net/emadeb/mzL6yjvx/18/

### Handsontable version

13.10

### Your environment

Chrome and Firefox

**Comment by adrianszymanski89** [comment]
Hi @emadebManyDesigns 

Thank you for reporting this. I was able to reproduce the issue with the larger dataset, as you mentioned, where the problem is visibly easily: https://jsfiddle.net/handsoncode/jdo0qu96/

I'll report this internally for further investigation and update you once we have a fix for this implemented.

**Comment by emadebManyDesigns** [comment]
Thanks.
For now I've fixed in the folowing way (but I'm not sure about the side effect):

core.js
```
  //use the changeSource 'validateBeforeApplyChanges' if empty
  this.setDataAtCell = function(row, column, value, source, applyChangesCallback) {
    ...
    validateChanges(processedChanges, changeSource ? changeSource : 'validateBeforeApplyChanges', isValid => {
      
  //when the source is 'validateBeforeApplyChanges' skip the cellRender
  this.validateCell = function(value, cellProperties, callback, source) {
    ...
    if (td && td.nodeName !== 'TH' && source !== 'validateBeforeApplyChanges') {
      const renderableRow = instance.rowIndexMapper.getRenderableFromVisualIndex(row);
      const renderableColumn = instance.columnIndexMapper.getRenderableFromVisualIndex(col);
      instance.view._wt.getSetting('cellRenderer', renderableRow, renderableColumn, td);
    }
```

**Comment by emadebManyDesigns** [comment]
I've noticed that another problem is that the beforeChange is called too much times.
https://jsfiddle.net/emadeb/mzL6yjvx/23/

In the new fiddle, if you open the console and you make a double click on a checkbox, 
you'll notice that the beforeChange is called 3 times instead of two: the last one has the old value equals to the new one.

![immagine](https://github.com/handsontable/handsontable/assets/41444394/76d7ab2e-7de5-48af-9b3a-cbd27e9afaa2)

**Comment by adrianszymanski89** [comment]
@emadebManyDesigns 

Thank you for more details and the workaround. It will surely be helpful for our develops during the investigation of the issue.

**Comment by AMBudnik** [comment]
Hi @emadebManyDesigns 

I am happy to announce that we have an update regarding this issue.

Here https://github.com/handsontable/handsontable/pull/10748 is a corresponding pull request that was merged and published in Handsontable v14.2. 

Could you please check if the newest version works for you?
